File: 2021/day07.py
import time

start = time.time()
with open("../inputs/2021/day07.input", "r") as f:
    inp = sorted(map(int, f.read().strip().split(",")))


def part1(inp):
    # Probably doesn't work for all inputs
    median = inp[len(inp) // 2]
    return sum(abs(k - median) for k in inp)


def part2(inp):
    # Probably doesn't work for all inputs
    mean = sum(inp) // len(inp)
    return sum((abs(k - mean) * (abs(k - mean) + 1)) // 2 for k in inp)


print(
    f"Day 07     {part1(inp):<14} {part2(inp):<14} {(time.time() - start)*1e3:>11.2f}"
)

File: rust-2021/src/day07.rs
use itertools::Itertools;

type T = i16;
type I = i32;

pub fn read(inp: &str) -> Vec<T> {
    inp.trim()
        .split(',')
        .map(|word| word.parse::<T>().unwrap())
        .sorted()
        .collect()
}

pub fn part1(inp: &[T]) -> I {
    let median = inp[inp.len() / 2];
    inp.iter()
        .fold(I::default(), |acc, i| acc + ((i - median).abs() as I))
}

pub fn part2(inp: &[T]) -> I {
    let mean: I = inp.iter().fold(I::default(), |acc, &i| acc + i as I) / (inp.len() as I);
    let mut ret = I::default();
    for &i in inp {
        let diff = (i as I - mean).abs();
        let gauss = diff * (diff + 1);
        ret += gauss;
    }
    ret >> 1
}

pub fn run() -> (String, String) {
    let inp = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day07.input"));
    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}
